### PR TITLE
feat[next]: Inline dynamic shifts

### PR DESCRIPTION
**Pull request by tehrengruber**
Dynamic shifts are not supported in the domain inference. In order to make them work nonetheless this PR aggressively inlines all arguments to `as_fieldop` until they contain only references to `itir.Program` params. Additionally the domain inference is extended to tolerate such `as_fieldop` by introducing a special domain marker that signifies a domain is unknown.

### CODE REVIEW

File: tests/next_tests/definitions.py
Comment (edopao) [inline tests/next_tests/definitions.py:155]: In order to see if DaCe GTIR backend supports dynamic offsets, we would have to delete line 155 (or I can wait for next merge commit, after #1648 is merged.

Comment (edopao) [reply tests/next_tests/definitions.py:155]: I tested it locally and it works.
image

**Comment by edopao** [comment]
I run icon4py stencils and I found that one small change is needed in dace backend. It makes sense to include it in this PR:
```
--- a/src/gt4py/next/program_processors/runners/dace_fieldview/gtir_dataflow.py
+++ b/src/gt4py/next/program_processors/runners/dace_fieldview/gtir_dataflow.py
@@ -1057,7 +1057,6 @@ class LambdaToDataflow(eve.NodeVisitor):
         """Implements shift in unstructured domain by means of a neighbor table."""
         assert connectivity.neighbor_axis in it.dimensions
         neighbor_dim = connectivity.neighbor_axis
-        assert neighbor_dim not in it.indices
 
         origin_dim = connectivity.origin_axis
         assert origin_dim in it.indices
```

**Comment by tehrengruber** [comment]
Feel free to push your Dace changes (as long as they are isolated to dace parts otherwise it'll be to confusing).

File: src/gt4py/next/iterator/transforms/fuse_as_fieldop.py
Comment (tehrengruber) [inline src/gt4py/next/iterator/transforms/fuse_as_fieldop.py:67]: All changes in this file are merely a small refactoring that extracts the field operator fusion for a single `as_fieldop` from the `FuseAsFieldOp` pass into a dedicated function `fuse_as_fieldop` used by the `InlineDynamicShifts` pass.

File: src/gt4py/next/iterator/transforms/infer_domain.py
Comment (tehrengruber) [inline src/gt4py/next/iterator/transforms/infer_domain.py:10]: @sf-n could you review the changes in this file?

**Review by SF-N — COMMENTED** [review]
Thanks, the changes make sense to me. The only thing one could reconsider is the name `allowed_unknown_domains` as it somehow implies to be a collection of domains rather than references.

File: src/gt4py/next/iterator/transforms/infer_domain.py
Comment (havogt) [inline src/gt4py/next/iterator/transforms/infer_domain.py:481]:
```suggestion
                stmt, offset_provider, symbolic_domain_sizes, allowed_unknown_domains=[param.id for param in program.params]
```
for readability

File: src/gt4py/next/iterator/transforms/infer_domain.py
Comment (havogt) [inline src/gt4py/next/iterator/transforms/infer_domain.py:429]: In general, I am confused why this is introduced... 
Are all program parameters allowed to be unknown, but the rest isn't?

Comment (havogt) [reply src/gt4py/next/iterator/transforms/infer_domain.py:429]: If I understand the logic, then I would remove it from here and rather write a separate path for checking this and avoid uglifying the this interface.

File: src/gt4py/next/iterator/transforms/inline_dynamic_shifts.py
Comment (havogt) [inline src/gt4py/next/iterator/transforms/inline_dynamic_shifts.py:32]: either 
```suggestion
            for offset in shifts[1::2]:
```
or use a function if you insist that does proper error checking.

File: src/gt4py/next/iterator/transforms/inline_dynamic_shifts.py
Comment (havogt) [inline src/gt4py/next/iterator/transforms/inline_dynamic_shifts.py:35]:
```suggestion
        dynamic_shifts = [any(trace_shifts.Sentinel.VALUE in shifts for shifts in param_shifts) for param_shifts in params_shifts]
```

File: src/gt4py/next/iterator/transforms/pass_manager.py
Comment (havogt) [inline src/gt4py/next/iterator/transforms/pass_manager.py:167]: Would it make sense to support unknown domain in the type inference for `as_offset`?

**Review by havogt — CHANGES_REQUESTED** [review]

File: src/gt4py/next/iterator/transforms/infer_domain.py
Comment (tehrengruber) [reply src/gt4py/next/iterator/transforms/infer_domain.py]: I renamed it `allow_uninferred` which now also allows if the domain is never accessed (previously indicated by `None` now `DomainAccessDescriptor.NEVER`)

File: src/gt4py/next/iterator/transforms/infer_domain.py
Comment (tehrengruber) [reply src/gt4py/next/iterator/transforms/infer_domain.py]: I introduced an enum `DomainAccessDescriptor` which indicates `UNKNOWN` or `NEVER` (previously `None`).

File: src/gt4py/next/iterator/transforms/infer_domain.py
Comment (tehrengruber) [reply src/gt4py/next/iterator/transforms/infer_domain.py]: I introduced a typed dict `InferenceOptions` that handles forwarding the options in all places where they are not used.

File: src/gt4py/next/iterator/transforms/pass_manager.py
Comment (tehrengruber) [reply src/gt4py/next/iterator/transforms/pass_manager.py:167]: I think this is a useful option for the domain inference and it naturally emerged after I addressed your review comments so I added it. I do however prefer working with an IR that has all domains inferred, which is the case after `InlineDynamicShifts` so I would leave it like that for now.

File: src/gt4py/next/iterator/transforms/infer_domain.py
Comment (havogt) [inline src/gt4py/next/iterator/transforms/infer_domain.py:33]:
```suggestion
    #: The access is unknown because of a dynamic shift whose extent is not known.
```

File: src/gt4py/next/iterator/transforms/infer_domain.py
Comment (havogt) [inline src/gt4py/next/iterator/transforms/infer_domain.py:86]:
```suggestion
        d  # type: ignore[misc]  # domain can never be unknown as these cases are filtered

File: src/gt4py/next/iterator/transforms/infer_domain.py
Comment (havogt) [inline src/gt4py/next/iterator/transforms/infer_domain.py:41]: please add a comment for the `tuple[DOMAIN]` case.

File: src/gt4py/next/iterator/transforms/infer_domain.py
Comment (havogt) [inline src/gt4py/next/iterator/transforms/infer_domain.py:158]:
```suggestion
    accessed_domains: ACCESSED_DOMAINS = {}
```

File: src/gt4py/next/iterator/transforms/infer_domain.py
Comment (havogt) [inline src/gt4py/next/iterator/transforms/infer_domain.py:181]:
```suggestion
    return accessed_domains
```

Comment (havogt) [reply src/gt4py/next/iterator/transforms/infer_domain.py:181]: please double check typing. It's unclear to me what it should be. Especially if `tuple[DOMAIN]` is included or not in the different cases.

File: src/gt4py/next/iterator/transforms/infer_domain.py
Comment (havogt) [inline src/gt4py/next/iterator/transforms/infer_domain.py:199]: What else can it be at this point?

File: src/gt4py/next/iterator/transforms/infer_domain.py
Comment (havogt) [reply src/gt4py/next/iterator/transforms/infer_domain.py:41]: or introduce a new alias for the one that includes tuple

Comment (havogt) [inline src/gt4py/next/iterator/transforms/infer_domain.py:41]:
```suggestion
_Domain: TypeAlias = domain_utils.SymbolicDomain | DomainAccessDescriptor | tuple["Domain", ...]
```

File: src/gt4py/next/iterator/transforms/infer_domain.py
Comment (havogt) [inline src/gt4py/next/iterator/transforms/infer_domain.py:42]:
```suggestion
_AccessedDomains: TypeAlias = dict[str, _Domain]
```
etc

File: src/gt4py/next/iterator/transforms/infer_domain.py
Comment (havogt) [inline src/gt4py/next/iterator/transforms/infer_domain.py:28]:
```suggestion
class _DomainAccessDescriptor(eve.StrEnum):
```
this is internal to the pass, correct?

File: src/gt4py/next/iterator/transforms/infer_domain.py
Comment (havogt) [inline src/gt4py/next/iterator/transforms/infer_domain.py:409]: "empty" is used only here, probably use a formulation with "never accessed" here

File: src/gt4py/next/iterator/transforms/infer_domain.py
Comment (havogt) [inline src/gt4py/next/iterator/transforms/infer_domain.py:30]: Add a note that this is only used internally for checking, then replaced by `None` (if I understand correctly).

**Review by havogt — CHANGES_REQUESTED** [review]
Mostly about clarifying types.

File: src/gt4py/next/iterator/transforms/infer_domain.py
Comment (tehrengruber) [reply src/gt4py/next/iterator/transforms/infer_domain.py]: Removed the cast, since it is not needed anymore.

File: src/gt4py/next/iterator/transforms/infer_domain.py
Comment (tehrengruber) [reply src/gt4py/next/iterator/transforms/infer_domain.py]: They can be used from the outside and this is also desired if you use `infer_expr` directly, so I left them public.

File: src/gt4py/next/iterator/transforms/infer_domain.py
Comment (tehrengruber) [reply src/gt4py/next/iterator/transforms/infer_domain.py]: See above.

File: src/gt4py/next/iterator/transforms/infer_domain.py
Comment (tehrengruber) [reply src/gt4py/next/iterator/transforms/infer_domain.py:28]: No, this can be used from the outside, it will for example appear in the `domain` annex attribute (used in `extract_tmp`) and could be used (I actually do this in the unused-return-from-scan fix I use in PMAP).

File: src/gt4py/next/iterator/transforms/infer_domain.py
Comment (tehrengruber) [reply src/gt4py/next/iterator/transforms/infer_domain.py:30]: This is not replaced by None, it appears in the `domain` annex attribute, but is not placed in the IR.

File: src/gt4py/next/iterator/transforms/infer_domain.py
Comment (havogt) [inline src/gt4py/next/iterator/transforms/infer_domain.py:47]: Can we make the TypeAliases private? Especially to avoid confusion with this `Domain` and `common.Domain`.

File: src/gt4py/next/iterator/transforms/infer_domain.py
Comment (havogt) [reply src/gt4py/next/iterator/transforms/infer_domain.py:30]: Do you need this information (the info what kind of "no domain" it is)? I don't really like that this ad-hoc concept leaks into other passes. Especially, since it is not clear how this Descriptor should be extended. If we call it `NoDomainDescriptor` at least it is clear there is no domain info.

File: src/gt4py/next/iterator/transforms/infer_domain.py
Comment (havogt) [reply src/gt4py/next/iterator/transforms/infer_domain.py]: Where do you do that?

File: src/gt4py/next/iterator/transforms/infer_domain.py
Comment (havogt) [inline src/gt4py/next/iterator/transforms/infer_domain.py:46]: I don't like `Domain`  here as it might create confusion with `common.Domain`.

File: src/gt4py/next/iterator/transforms/infer_domain.py
Comment (tehrengruber) [reply src/gt4py/next/iterator/transforms/infer_domain.py]: Here: https://github.com/GridTools/gt4py/pull/1738/files#diff-36c97ae1e7197fbdd81532e533a2e6294febc1c23bbb1a6ef2968130e15368fbR77
I added an annotation there so that we can reverse search.

File: src/gt4py/next/iterator/transforms/infer_domain.py
Comment (tehrengruber) [reply src/gt4py/next/iterator/transforms/infer_domain.py:30]: TODO: write that DomainAccessDescriptor.KNOWN is missing, but skipped since we don't need it for technical reasons.

File: src/gt4py/next/iterator/transforms/infer_domain.py
Comment (tehrengruber) [reply src/gt4py/next/iterator/transforms/infer_domain.py]: Renamed to `DomainAccess` as discussed.

**Review by havogt — APPROVED** [review]
lgtm

**Pull request merged**
